searchcore/src/vespa/searchcore/proton/server/buckethandler.cpp

```diff
@@ -147,9 +148,20 @@ void
 BucketHandler::notifyClusterStateChanged(const std::shared_ptr<IBucketStateCalculator> & newCalc)
 {
     bool oldNodeUp = _nodeUp;
-    _nodeUp = newCalc->nodeUp();
+    bool oldNodeMaintenance = _nodeMaintenance;
+    _nodeUp = newCalc->nodeUp(); // Up, Retired or Initializing
+    _nodeMaintenance = newCalc->nodeMaintenance();
     LOG(spam, "notifyClusterStateChanged: %s -> %s", oldNodeUp ? "up" : "down", _nodeUp ? "up" : "down");
-    if (oldNodeUp && !_nodeUp) {
+    if (_nodeMaintenance) {
+        return; // Don't deactivate buckets in maintenance mode; let query traffic drain away naturally.
+    }
+    // We implicitly deactivate buckets in two edge cases:
+    //  - Up -> Down (not maintenance; handled above), since the node can not be expected to offer
+    //    any graceful query draining when set Down.
+    //  - Maintenance -> Up, since we'd otherwise introduce a bunch of transient duplicate results
+    //    into queries. The assumption is that the system has already activated buckets on other
+    //    nodes in such a scenario.
+    if ((oldNodeUp && !_nodeUp) || (oldNodeMaintenance && _nodeUp)) {
```

Comment:
The transitions up -> maintenance -> down -> up seems to not trigger deactivation of buckets.
If old state is maintenance and new state is not maintenance then all buckets should be deactivated.

i.e. instead of `(oldNodeMaintenance && _nodeUp)` consider using `(oldNodeMaintenance && !_nodeMaintenance)` or `(!oldNodeUp && _nodeUp)`

Reply:
Good point. I think `oldNodeMaintenance && !_nodeMaintenance` might be the best, since this particular edge case isn't expected to happen except from manual administrative tasks (i.e. `vespa-set-node-state`). `(!oldNodeUp && _nodeUp)` would presumably trigger on every normal startup, when everything is already implicitly marked as inactive.

However, I think I need to ponder these changes a bit more since the current interaction with having multiple bucket spaces with differing maintenance state (i.e. global document merging) isn't as well-defined as I would like it to be.